sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,300 @@
+                                         "for label, %s, does not conform "
+                                         "to the number of features, %d"
+                                         % (class_key, n_features))
+                elif not isinstance(val, (np.floating, float,
+                                          numbers.Integral, np.integer)):
+                    raise ValueError("If missing_proba is a dict with "
+                                     "target labels as keys, the values of "
+                                     "the dict should either be a single "
+                                     "float or an array of shape "
+                                     "(n_features,). %r was passed for class "
+                                     "label %s" % (val, class_key))
+
+                drop_probs[encoded_class_key, :] = val
+
+        else:
+            # For MCAR
+            # Validate and convert the missing_proba dict into a
+            # 1D probability distribution along the features
+
+            drop_probs = np.zeros((1, n_features), dtype=np.float64)
+
+            if isinstance(self.missing_proba, (list, tuple, np.ndarray)):
+                # Convert to ndarray and check shape
+                missing_proba = np.asarray(self.missing_proba)
+                if missing_proba.shape[0] != n_features:
+                    raise ValueError("The shape of the per feature "
+                                     "drop-probabilities vector does not "
+                                     "conform to the number of features, %d"
+                                     % n_features)
+            elif not isinstance(self.missing_proba,
+                                (np.floating, float, numbers.Integral,
+                                 np.integer)):
+                raise ValueError("missing_proba must be a float or "
+                                 "1D vector (list, tuple or np.ndarray) of "
+                                 "shape (n_features,) or dict of 1D vector / "
+                                 "floats. %r was passed"
+                                 % self.missing_proba)
+
+            drop_probs[:] = self.missing_proba
+            # Hack to simplify code
+            classes = [0, ]
+            y = np.zeros(n_samples)
+
+        if np.any(drop_probs < 0) or np.any(drop_probs > 1):
+            raise ValueError("All the individual drop-probabilities should be "
+                             "within the range of [0, 1]. The given "
+                             "missing_proba does not conform to that. %r"
+                             % self.missing_proba)
+
+        for i, class_i in enumerate(classes):
+            samples_mask = (y == class_i)
+            this_n_samples = samples_mask.sum()
+            this_block_indices = np.arange(n_samples)[samples_mask]
+
+            for feature in range(n_features):
+                # Shuffle even if this_required_n_missing is 0, to maintain
```

Comment:
This is very expensive when there are many features, no? Would it be more sensible to just generate a starting `random_state` for each class / feature?